### ISSUE TITLE
Improved `count` support

### ISSUE THREAD

**Issue by mawkler** [kylechui/nvim-surround#147]
**Checklist**

- [x] Have you read through `:h nvim-surround` to see if there might be any relevant information there?

**Is your feature request related to a problem? Please describe.**

I think it would be really cool if I for instance could do something like `2siw*` to make a word bold in markdown (i.e. convert `word` into `**word**`), or do the opposite: `2ds*` to make text non-bold in markdown (i.e. convert `**word**` into `word`).

**Describe the solution you'd like**

Prepending a count (i.e. a number 1-9) before a surround command performs the surround that many times over the same text-object.

**Comment by smjonas** [comment]
I've had the same feature request here: https://github.com/kylechui/nvim-surround/issues/86

There might be some relevant info there :) Btw, I still think this feature is useful (and there's no workaround for your usecase, e.g. ds2* doesn't work).

**Comment by kylechui** [comment]
@smjonas It seems like this is slightly different; your feature request was about surrounding larger selections (e.g. two words at a time), whereas this seems to be about surrounding a single word *twice*. @melkster This sounds like a good idea; unfortunately at the moment I need to first get through the other updates (*good* query support) and some other issues, and then learn how counting works. Apologies in advance if this takes a while, as I'll have less time to work on this during the school year/internship application season.

**Comment by kylechui** [comment]
Hello! It's been quite a while (sorry), but I'm productively procrastinating on my schoolwork by working on this plugin :smile: Coming back to this issue, it seems that we kind of have two different ways to approach this problem: prefixing the operator or prefixing the surround, i.e. `2ysiw*` vs. `ysiw2*`. My current thoughts are that the former has no real disadvantages (and can probably be implemented by leveraging the built-in `v:count`), while the latter would require me to:
* Parse numbers myself to see how many times to repeat a given action (seems relatively easy IMO)
* Force a breaking change where numbers are no longer valid surround keys (bigger deal)

**Comment by kylechui** [comment]
I just realized that there's already a bit of inconsistency with how things work for the `yss` special case. As of right now, `2yss)` would surround two lines with parentheses, instead of surrounding the current line with two sets of parentheses. I'm not exactly sure how to reconcile these differences; if anybody has any ideas feel free to let me know!

**Comment by kylechui** [comment]
Continuing the tradition of replying once a year (sorry!). It seems like I was mistaken before, and that prefixing the motion with a count already applies the count to the motion, i.e. `2ysaw)` will surround two words with `(...)`. Thus `2yss)` really isn't a "special case", which makes me even less inclined to change the behavior now. The only real option seems to be something like `ysiw10)`, but like I mentioned earlier, I'm not sure if I want to implement the parsing for that...

**Comment by kylechui** [comment]
I think I'll close this issue for now since implementing this seems to mandate some kind of breaking change, but if anybody has any new ideas that haven't already been posted in this thread, feel free to comment!

Tangentially related is that with the `"sticky"` option for `move_cursor` (soon-to-be-merged in #334), one can just dot-repeat the surround to get multiple.

**Comment by kylechui** [comment]
I haven't solved the issue from a design standpoint (i.e. how to actually get the inputs in), but another person asked about this recently so I'm just saving this implementation snippet here for posterity:
```
+-- Repeats a delimiter pair n times.
+---@param delimiters delimiter_pair The delimiters to be repeated.
+---@param n integer The number of times to repeat the delimiters.
+---@nodiscard
+M.repeat_delimiters = function(delimiters, n)
+    local acc = { { "" }, { "" } }
+    for _ = 1, n do
+        acc[1][#acc[1]] = acc[1][#acc[1]] .. delimiters[1][1]
+        vim.list_extend(acc[1], delimiters[1], 2)
+        acc[2][#acc[2]] = acc[2][#acc[2]] .. delimiters[2][1]
+        vim.list_extend(acc[2], delimiters[2], 2)
+    end
+    return acc
+end
```

**Comment by kylechui** [comment]
I'm starting to get sold on the idea that `ysiw2*` on `hello` ->` `**hello**` is a good idea; might try implementing it in a feature branch :thinking:

**Comment by mawkler** [comment]
@kylechui

> ~~As mentioned before, it's ambiguous if this "should" be `((hello)) world` or `(hello world)`~~. The correct behavior here is (hello world)

I personally disagree. To me as a user it makes most sense if `2ysiw)` results in `((hello))`, since the count is closest to the `ys`. If the user wants `(hello world)` they should do `ys2iw)`.

Moreover, `ysiw2)` introduces a new ambiguity if the user wants to surround something with a number like `2hello2`, or if they have created their custom surround that's mapped to a number. Just as an arbitrary example they might want `ysiw2` to surround the text with a level 2 heading (`subsection{hello}`) in LaTeX.

> prefixing a surround with a count caused the count to be used both for the motion and the repeat count

I do however understand if you decide to go with `ysiw2)` from an implementation perspective if it's impossible to implement the former.

**Comment by kylechui** [comment]
Thanks for weighing in! I think the prefix count and middle count get multiplied to form the motion, e.g. `2d3w` == `d6w` == `6dw`. I actually tried implementing `2ysiw)`, however the motion always defaulted to selecting `2iw`. I agree that it makes more sense for the prefix count to apply to the number of surrounds, rather than the size of the motion.

As for the suffix count, e.g. `ysiw2)`---you are right that it is incompatible with number surrounds, and I was considering making a breaking change for this. Actually, either way this will be a breaking change because we have to choose which set of semantics is right:
- No number surrounds (breaking!), `yss2b` surrounds 1 line with 2 sets of parentheses (if I can figure this out), 2yssb surrounds 2 lines with 1 set of parentheses
- Keep number surrounds, 2yssb surrounds 1 line with 2 sets of parentheses (breaking!)

**Comment by kylechui** [comment]
Hello again, I think I've decided on the syntax: `[count?]ys[motion][char]`, so something like `2ys3w)` will surround 3 words with `((`, `))`. Some preliminary work has been done in #360

**Comment by mawkler** [comment]
@kylechui Thank you so much for implementing this! I tried the PR and found no issues.

> I still need to think about what prefixing `ds` and `cs` should do, or if that's even a feature that people want to see

For `ds`, to me it makes sense if `2ds*` does the opposite to `2ysiw*`, i.e. it would turn `**foo**` into `foo`.

As for `cs` I see two possible behaviours:

1. The count states how many surrounding characters the new surround should have. E.g. given the text `(foo bar)`, doing `3cs([` would yield `[[[foo bar]]]`.
2. The count states how many duplicates of the surrounding character pair to change. E.g. given the text `**foo bar**`, doing `2cs*_` would yield `_foo bar_`.

I personally think that 1 is the more useful one.

**Comment by kylechui** [comment]
I would agree that 1 is more useful; perhaps I should merge in the current progress first and work on delete/change later? One case I'm thinking about is:
```
((foo) bar (baz))```
If the cursor is on `bar`, should `2ds)` delete the  one before xor after, in addition to the outermost pair? Or should it depend on the users configuration, e.g. `sticky` vs. `first` cursor?

**Comment by mawkler** [comment]
@kylechui 

> perhaps I should merge in the current progress first and work on delete/change later?

That sounds like a good idea!

> ```
> ((foo) bar (baz))
> ```

I'm thinking that the optimal behaviour is that `2ds)` behaves the same way that `ds).` would have behaved.

**Comment by kylechui** [comment]
Woops didn't mean to close this quite yet; still need to implement the delete/change operations

**Comment by kylechui** [comment]
Deleting seems to work just fine; the main issue I'm running into right now is that for `change` operations, figuring out how to highlight the stuff-to-change seems very hard if not impossible. The delete code very naively just replays the `delete` action `vim.v.count1` times. However, for `change` operation, if we wish to highlight all of the delimiters-to-be-changed, we need to know a priori _where_ everything is _before we make any changes_.

**Comment by kylechui** [comment]
OK I think that repeating delete/change operations should mostly work on the `feat/repeat-delete-change` branch. It seems to work based on my own testing + passes all of my tests, but I would still like others to play around with it for a little bit before I release it as `v3.1.0`, as I've done a bit of refactoring in between. The only caveat that I can find right now is that the plugin will only highlight the first surround during a `change` operation, instead of up to N surrounding delimiter pairs. However, I do not think that should be a blocker for release since it is just a cosmetic issue. Thanks in advance for any feedback and in general for using this plugin!

**Comment by mawkler** [comment]
@kylechui I gave `feat/repeat-delete-change` a try and found no issues. I notice that you went with foo nr 2 out of the two suggestions in [my previous comment](https://github.com/kylechui/nvim-surround/issues/147#issuecomment-2709999697). May I ask why? 🙂

**Comment by kylechui** [comment]
Better late than never, right? 😅
